File: packages/shared-types/src/service-behavior.test.ts
import { describe, expect, it } from 'vitest';
import { ServiceBehaviorSchema } from './service-behavior';

function textField(key = 'room') {
  return {
    key,
    label: { en: 'Room', ru: 'Кабинет' },
    type: 'text' as const,
    required: true
  };
}

function groupAtDepth(depth: number): unknown {
  let node: unknown = {
    kind: 'rule',
    field: 'identity.isAuthenticated',
    operator: 'is-true'
  };
  for (let index = 0; index < depth; index += 1) {
    node = { kind: 'group', combinator: 'and', children: [node] };
  }
  return node;
}

describe('service behavior schema', () => {
  it('parses portable keyed options and the canonical Task 2 access policy', () => {
    const behavior = ServiceBehaviorSchema.parse({
      version: 1,
      information: {
        body: { en: 'Bring an employee badge', ru: 'Возьмите пропуск' },
        requireAcknowledgement: true
      },
      fields: [
        textField(),
        {
          key: 'floor',
          label: { en: 'Floor', ru: 'Этаж' },
          type: 'number',
          required: false
        },
        {
          key: 'contact_phone',
          label: { en: 'Phone', ru: 'Телефон' },
          type: 'phone',
          required: false
        },
        {
          key: 'visit_reason',
          label: { en: 'Reason', ru: 'Причина' },
          type: 'select',
          required: true,
          options: [
            {
              key: 'consultation',
              label: { en: 'Consultation', ru: 'Консультация' }
            },
            { key: 'pickup', label: { en: 'Pickup', ru: 'Получение' } }
          ]
        },
        {
          key: 'arrival',
          label: { en: 'Arrival', ru: 'Прибыл' },
          type: 'checkbox',
          required: true
        }
      ],
      dataRetentionDays: 7,
      route: { mode: 'page-slot', slot: 'service-form' },
      access: {
        when: {
          kind: 'rule',
          field: 'identity.isAuthenticated',
          operator: 'is-true'
        },
        whenFalse: 'lock'
      }
    });

    const select = behavior.fields[3];
    if (select?.type !== 'select') {
      throw new Error('expected select field');
    }
    expect(select.options[0]?.key).toBe('consultation');
    expect(behavior.access?.whenFalse).toBe('lock');
  });

  it.each([
    {
      name: 'an unsupported behavior version',
      behavior: { version: 2 }
    },
    {
      name: 'a field key outside the portable key format',
      behavior: {
        version: 1,
        fields: [{ ...textField('Room') }],
        dataRetentionDays: 1
      }
    },
    {
      name: 'duplicate requested field keys',
      behavior: {
        version: 1,
        fields: [textField('room'), { ...textField('room'), type: 'number' }],
        dataRetentionDays: 1
      }
    },
    {
      name: 'requested fields without retention',
      behavior: { version: 1, fields: [textField()] }
    },
    {
      name: 'select options without stable keys',
      behavior: {
        version: 1,
        fields: [
          {
            key: 'reason',
            label: { en: 'Reason' },
            type: 'select',
            required: true,
            options: [{ en: 'Consultation' }]
          }
        ],
        dataRetentionDays: 1
      }
    },
    {
      name: 'duplicate select option keys',
      behavior: {
        version: 1,
        fields: [
          {
            key: 'reason',
            label: { en: 'Reason' },
            type: 'select',
            required: true,
            options: [
              { key: 'consultation', label: { en: 'Consultation' } },
              { key: 'consultation', label: { en: 'Second consultation' } }
            ]
          }
        ],
        dataRetentionDays: 1
      }
    },
    {
      name: 'an access rule with an incompatible operator',
      behavior: {
        version: 1,
        access: {
          when: {
            kind: 'rule',
            field: 'identity.groups',
            operator: 'gt',
            value: 1
          },
          whenFalse: 'hide'
        }
      }
    },
    {
      name: 'a route outside the page-slot union',
      behavior: {
        version: 1,
        route: { mode: 'page-slot', slot: 'service-picker' }
      }
    }
  ])('rejects $name', ({ behavior }) => {
    expect(ServiceBehaviorSchema.safeParse(behavior).success).toBe(false);
  });

  it.each([
    [
      'more than 50 fields',
      {
        version: 1,
        fields: Array.from({ length: 51 }, (_, index) =>
          textField(`field_${index}`)
        ),
        dataRetentionDays: 1
      }
    ],
    [
      'more than eight localized entries',
      {
        version: 1,
        fields: [
          {
            ...textField(),
            label: Object.fromEntries(
              Array.from({ length: 9 }, (_, index) => [`l${index}`, 'Room'])
            )
          }
        ],
        dataRetentionDays: 1
      }
    ],
    [
      'a locale key longer than 16 characters',
      {
        version: 1,
        fields: [
          { ...textField(), label: { this_locale_key_is_too_long: 'Room' } }
        ],
        dataRetentionDays: 1
      }
    ],
    [
      'a field label longer than 160 characters',
      {
        version: 1,
        fields: [{ ...textField(), label: { en: 'a'.repeat(161) } }],
        dataRetentionDays: 1
      }
    ],
    [
      'an information body longer than 4000 characters',
      { version: 1, information: { body: { en: 'a'.repeat(4001) } } }
    ],
    [
      'more than 50 select options',
      {
        version: 1,
        fields: [
          {
            key: 'reason',
            label: { en: 'Reason' },
            type: 'select',
            required: true,
            options: Array.from({ length: 51 }, (_, index) => ({
              key: `option_${index}`,
              label: { en: 'Option' }
            }))
          }
        ],
        dataRetentionDays: 1
      }
    ],
    [
      'an access AST deeper than eight groups',
      { version: 1, access: { when: groupAtDepth(9), whenFalse: 'hide' } }
    ],
    [
      'more than 100 access AST nodes',
      {
        version: 1,
        access: {
          when: {
            kind: 'group',
            combinator: 'or',
            children: Array.from({ length: 20 }, () => ({
              kind: 'group',
              combinator: 'and',
              children: Array.from({ length: 4 }, () => ({
                kind: 'rule',
                field: 'live.isOpen',
                operator: 'is-true'
              }))
            }))
          },
          whenFalse: 'hide'
        }
      }
    ],
    [
      'an access group with more than 20 children',
      {
        version: 1,
        access: {
          when: {
            kind: 'group',
            combinator: 'or',
            children: Array.from({ length: 21 }, () => ({
              kind: 'rule',
              field: 'live.isOpen',
              operator: 'is-true'
            }))
          },
          whenFalse: 'hide'
        }
      }
    ]
  ])('rejects $0', (_name, behavior) => {
    expect(ServiceBehaviorSchema.safeParse(behavior).success).toBe(false);
  });

  it('accepts the access AST boundaries and rejects raw behavior above 64 KiB', () => {
    const oneHundredNodes = Array.from({ length: 20 }, (_, groupIndex) => ({
      kind: 'group' as const,
      combinator: 'or' as const,
      children: Array.from({ length: groupIndex === 19 ? 3 : 4 }, () => ({
        kind: 'rule' as const,
        field: 'live.isOpen' as const,
        operator: 'is-true' as const
      }))
    }));
    const atBound = {
      version: 1,
      access: {
        when: {
          kind: 'group' as const,
          combinator: 'or' as const,
          children: oneHundredNodes
        },
        whenFalse: 'hide' as const
      }
    };
    const localized = Object.fromEntries(
      Array.from({ length: 8 }, (_, index) => [`l${index}`, 'x'.repeat(160)])
    );
    const atSizeBase = {
      version: 1,
      information: { body: { en: 'x' } },
      route: { mode: 'auto' as const },
      access: {
        when: {
          kind: 'rule' as const,
          field: 'live.isOpen' as const,
          operator: 'is-true' as const
        },
        whenFalse: 'hide' as const
      },
      fields: Array.from({ length: 22 }, (_, index) => ({
        key: `field_${index}`,
        label: localized,
        type: 'select' as const,
        required: false,
        options: [{ key: 'option', label: localized }]
      })),
      dataRetentionDays: 1
    };
    const baseBytes = new TextEncoder().encode(
      JSON.stringify(atSizeBase)
    ).length;
    const padding = 64 * 1024 - baseBytes;
    const atSize = {
      ...atSizeBase,
      information: { body: { en: 'x'.repeat(padding + 1) } }
    };
    const aboveSize = {
      ...atSize,
      information: { body: { en: 'x'.repeat(padding + 2) } }
    };

    expect(padding).toBeGreaterThanOrEqual(0);
    expect(padding + 1).toBeLessThanOrEqual(4000);
    expect(new TextEncoder().encode(JSON.stringify(atSize)).length).toBe(
      64 * 1024
    );
    expect(
      ServiceBehaviorSchema.safeParse({
        version: 1,
        access: { when: groupAtDepth(7), whenFalse: 'hide' }
      }).success
    ).toBe(true);
    expect(ServiceBehaviorSchema.safeParse(atBound).success).toBe(true);
    expect(ServiceBehaviorSchema.safeParse(atSize).success).toBe(true);
    expect(ServiceBehaviorSchema.safeParse(aboveSize).success).toBe(false);
  });

  it('rejects records with inherited required behavior keys at every behavior boundary', () => {
    const inheritedBehavior = Object.create({
      version: 1,
      fields: [],
      route: { mode: 'auto' }
    });
    const inheritedLabel = Object.create({ en: 'Room' });
    const inheritedOption = Object.create({
      key: 'consultation',
      label: { en: 'Consultation' }
    });

    expect(ServiceBehaviorSchema.safeParse(inheritedBehavior).success).toBe(
      false
    );
    expect(
      ServiceBehaviorSchema.safeParse({
        version: 1,
        fields: [{ ...textField(), label: inheritedLabel }],
        dataRetentionDays: 1
      }).success
    ).toBe(false);
    expect(
      ServiceBehaviorSchema.safeParse({
        version: 1,
        fields: [
          {
            key: 'reason',
            label: { en: 'Reason' },
            type: 'select',
            required: true,
            options: [inheritedOption]
          }
        ],
        dataRetentionDays: 1
      }).success
    ).toBe(false);
  });
});

File: apps/backend/internal/services/service_behavior_test.go
package services

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"testing"

	"quokkaq-go-backend/internal/models"

	"gorm.io/gorm"
)

const validServiceBehaviorJSON = `{
  "version": 1,
  "information": {"body": {"en": "Bring your badge"}, "requireAcknowledgement": true},
  "fields": [
    {"key": "room", "label": {"en": "Room"}, "type": "text", "required": true},
    {"key": "reason", "label": {"en": "Reason"}, "type": "select", "required": false, "options": [{"key":"consultation","label":{"en":"Consultation"}}]}
  ],
  "dataRetentionDays": 7,
  "route": {"mode": "page-slot", "slot": "service-form"},
  "access": {"when": {"kind": "rule", "field": "identity.isAuthenticated", "operator": "is-true"}, "whenFalse": "lock"}
}`

func TestValidateServiceBehaviorJSON(t *testing.T) {
	tests := []struct {
		name string
		raw  string
		want error
	}{
		{name: "absent behavior", raw: "", want: nil},
		{name: "JSON null behavior", raw: "null", want: nil},
		{name: "invalid json", raw: `{`, want: ErrServiceBehaviorInvalid},
		{name: "unsupported version", raw: `{"version":2}`, want: ErrServiceBehaviorInvalid},
		{name: "duplicate field key", raw: `{"version":1,"fields":[{"key":"room","label":{"en":"Room"},"type":"text","required":true},{"key":"room","label":{"en":"Second room"},"type":"number","required":false}],"dataRetentionDays":1}`, want: ErrServiceBehaviorInvalid},
		{name: "select without options", raw: `{"version":1,"fields":[{"key":"reason","label":{"en":"Reason"},"type":"select","required":true}],"dataRetentionDays":1}`, want: ErrServiceBehaviorInvalid},
		{name: "select option without stable key", raw: `{"version":1,"fields":[{"key":"reason","label":{"en":"Reason"},"type":"select","required":true,"options":[{"en":"Consultation"}]}],"dataRetentionDays":1}`, want: ErrServiceBehaviorInvalid},
		{name: "duplicate select option key", raw: `{"version":1,"fields":[{"key":"reason","label":{"en":"Reason"},"type":"select","required":true,"options":[{"key":"consultation","label":{"en":"Consultation"}},{"key":"consultation","label":{"en":"Again"}}]}],"dataRetentionDays":1}`, want: ErrServiceBehaviorInvalid},
		{name: "incompatible condition operator", raw: `{"version":1,"access":{"when":{"kind":"rule","field":"identity.groups","operator":"gt","value":1},"whenFalse":"lock"}}`, want: ErrServiceBehaviorInvalid},
		{name: "null numeric condition value", raw: `{"version":1,"access":{"when":{"kind":"rule","field":"live.queueLength","operator":"gt","value":null},"whenFalse":"lock"}}`, want: ErrServiceBehaviorInvalid},
		{name: "canonical integral JSON forms", raw: `{"version":1e0,"fields":[{"key":"room","label":{"en":"Room"},"type":"text","required":true}],"dataRetentionDays":1.0}`, want: nil},
		{name: "fractional retention", raw: `{"version":1,"fields":[{"key":"room","label":{"en":"Room"},"type":"text","required":true}],"dataRetentionDays":1.5}`, want: ErrServiceBehaviorInvalid},
		{name: "valid behavior", raw: validServiceBehaviorJSON, want: nil},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateServiceBehaviorJSON(json.RawMessage(tt.raw))
			if !errors.Is(err, tt.want) {
				t.Fatalf("ValidateServiceBehaviorJSON() error = %v, want %v", err, tt.want)
			}
		})
	}
}

func TestServiceBehaviorModelDecodingAcceptsIntegralJSONForms(t *testing.T) {
	var service models.Service
	if err := json.Unmarshal([]byte(`{"unitId":"unit-1","behavior":{"version":1e0,"fields":[{"key":"room","label":{"en":"Room"},"type":"text","required":true}],"dataRetentionDays":1.0}}`), &service); err != nil {
		t.Fatalf("service JSON decode rejected integral behavior numbers: %v", err)
	}
	if err := ValidateServiceBehaviorJSON(service.Behavior.RawJSON()); err != nil {
		t.Fatalf("decoded behavior failed validation: %v", err)
	}
}

func TestValidateServiceBehaviorJSON_resourceBounds(t *testing.T) {
	validField := func(key string) map[string]any {
		return map[string]any{"key": key, "label": map[string]any{"en": "Room"}, "type": "text", "required": false}
	}
	fields := make([]any, 51)
	for i := range fields {
		fields[i] = validField(fmt.Sprintf("field_%d", i))
	}
	localizedNine := map[string]any{}
	for i := 0; i < 9; i++ {
		localizedNine[fmt.Sprintf("l%d", i)] = "Room"
	}
	tooManyOptions := make([]any, 51)
	for i := range tooManyOptions {
		tooManyOptions[i] = map[string]any{"key": fmt.Sprintf("option_%d", i), "label": map[string]any{"en": "Option"}}
	}
	deep := conditionGroupAtDepth(9)
	tooManyChildren := make([]any, 21)
	for i := range tooManyChildren {
		tooManyChildren[i] = map[string]any{"kind": "rule", "field": "live.isOpen", "operator": "is-true"}
	}
	tooManyNodes := make([]any, 20)
	for groupIndex := range tooManyNodes {
		rules := make([]any, 4)
		for ruleIndex := range rules {
			rules[ruleIndex] = map[string]any{"kind": "rule", "field": "live.isOpen", "operator": "is-true"}
		}
		tooManyNodes[groupIndex] = map[string]any{"kind": "group", "combinator": "or", "children": rules}
	}

	tests := []struct {
		name     string
		behavior map[string]any
		want     error
	}{
		{name: "at field bound", behavior: map[string]any{"version": 1, "fields": []any{validField("room")}, "dataRetentionDays": 1}, want: nil},
		{name: "more than 50 fields", behavior: map[string]any{"version": 1, "fields": fields, "dataRetentionDays": 1}, want: ErrServiceBehaviorInvalid},
		{name: "more than eight localized entries", behavior: map[string]any{"version": 1, "fields": []any{map[string]any{"key": "room", "label": localizedNine, "type": "text", "required": true}}, "dataRetentionDays": 1}, want: ErrServiceBehaviorInvalid},
		{name: "locale key longer than sixteen", behavior: map[string]any{"version": 1, "fields": []any{map[string]any{"key": "room", "label": map[string]any{"this_locale_key_is_too_long": "Room"}, "type": "text", "required": true}}, "dataRetentionDays": 1}, want: ErrServiceBehaviorInvalid},
		{name: "field label longer than 160", behavior: map[string]any{"version": 1, "fields": []any{map[string]any{"key": "room", "label": map[string]any{"en": strings.Repeat("a", 161)}, "type": "text", "required": true}}, "dataRetentionDays": 1}, want: ErrServiceBehaviorInvalid},
		{name: "information body longer than 4000", behavior: map[string]any{"version": 1, "information": map[string]any{"body": map[string]any{"en": strings.Repeat("a", 4001)}}, "fields": []any{}}, want: ErrServiceBehaviorInvalid},
		{name: "more than 50 select options", behavior: map[string]any{"version": 1, "fields": []any{map[string]any{"key": "reason", "label": map[string]any{"en": "Reason"}, "type": "select", "required": false, "options": tooManyOptions}}, "dataRetentionDays": 1}, want: ErrServiceBehaviorInvalid},
		{name: "access deeper than eight nodes", behavior: map[string]any{"version": 1, "access": map[string]any{"when": deep, "whenFalse": "hide"}}, want: ErrServiceBehaviorInvalid},
		{name: "access at depth eight", behavior: map[string]any{"version": 1, "access": map[string]any{"when": conditionGroupAtDepth(7), "whenFalse": "hide"}}, want: nil},
		{name: "access with more than 100 nodes", behavior: map[string]any{"version": 1, "access": map[string]any{"when": map[string]any{"kind": "group", "combinator": "or", "children": tooManyNodes}, "whenFalse": "hide"}}, want: ErrServiceBehaviorInvalid},
		{name: "access group with more than 20 children", behavior: map[string]any{"version": 1, "access": map[string]any{"when": map[string]any{"kind": "group", "combinator": "and", "children": tooManyChildren}, "whenFalse": "hide"}}, want: ErrServiceBehaviorInvalid},
		{name: "access at 100 nodes", behavior: map[string]any{"version": 1, "access": map[string]any{"when": conditionAtOneHundredNodes(), "whenFalse": "hide"}}, want: nil},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			raw, err := json.Marshal(tt.behavior)
			if err != nil {
				t.Fatal(err)
			}
			err = ValidateServiceBehaviorJSON(raw)
			if !errors.Is(err, tt.want) {
				t.Fatalf("ValidateServiceBehaviorJSON() error = %v, want %v", err, tt.want)
			}
		})
	}

	atLimit := serviceBehaviorJSONAtSize(t, 64*1024)
	if len(atLimit) != 64*1024 {
		t.Fatalf("atLimit length = %d", len(atLimit))
	}
	if err := ValidateServiceBehaviorJSON(atLimit); err != nil {
		t.Fatalf("at-limit behavior rejected: %v", err)
	}
	aboveLimit := serviceBehaviorJSONAtSize(t, 64*1024+1)
	if err := ValidateServiceBehaviorJSON(aboveLimit); !errors.Is(err, ErrServiceBehaviorInvalid) {
		t.Fatalf("above-limit behavior error = %v", err)
	}
}

func conditionGroupAtDepth(groups int) map[string]any {
	var node any = map[string]any{"kind": "rule", "field": "live.isOpen", "operator": "is-true"}
	for i := 0; i < groups; i++ {
		node = map[string]any{"kind": "group", "combinator": "and", "children": []any{node}}
	}
	return node.(map[string]any)
}

func conditionAtOneHundredNodes() map[string]any {
	children := make([]any, 20)
	for groupIndex := range children {
		ruleCount := 4
		if groupIndex == 19 {
			ruleCount = 3
		}
		rules := make([]any, ruleCount)
		for ruleIndex := range rules {
			rules[ruleIndex] = map[string]any{"kind": "rule", "field": "live.isOpen", "operator": "is-true"}
		}
		children[groupIndex] = map[string]any{"kind": "group", "combinator": "or", "children": rules}
	}
	return map[string]any{"kind": "group", "combinator": "or", "children": children}
}

func serviceBehaviorJSONAtSize(t *testing.T, size int) json.RawMessage {
	t.Helper()
	localized := map[string]any{}
	for i := 0; i < 8; i++ {
		localized[fmt.Sprintf("l%d", i)] = strings.Repeat("x", 160)
	}
	fields := make([]any, 22)
	for i := range fields {
		fields[i] = map[string]any{
			"key":      fmt.Sprintf("field_%d", i),
			"label":    localized,
			"type":     "select",
			"required": false,
			"options":  []any{map[string]any{"key": "option", "label": localized}},
		}
	}
	behavior := map[string]any{
		"version":     1,
		"information": map[string]any{"body": map[string]any{"en": "x"}},
		"route":       map[string]any{"mode": "auto"},
		"access": map[string]any{
			"when":      map[string]any{"kind": "rule", "field": "live.isOpen", "operator": "is-true"},
			"whenFalse": "hide",
		},
		"fields":            fields,
		"dataRetentionDays": 1,
	}
	base, err := json.Marshal(behavior)
	if err != nil {
		t.Fatal(err)
	}
	padding := size - len(base)
	if padding < 0 || padding+1 > 4000 {
		t.Fatalf("cannot construct %d-byte behavior; padding=%d", size, padding)
	}
	behavior["information"] = map[string]any{"body": map[string]any{"en": strings.Repeat("x", padding+1)}}
	out, err := json.Marshal(behavior)
	if err != nil {
		t.Fatal(err)
	}
	return out
}

func TestServiceBehaviorValidationRunsBeforeCreateAndUpdatePersistence(t *testing.T) {
	invalid, err := models.ParseServiceBehaviorJSON(json.RawMessage(`{"version":2}`))
	if err != nil {
		t.Fatal(err)
	}

	t.Run("create", func(t *testing.T) {
		service := &models.Service{UnitID: "unit-1", Behavior: invalid}
		err := NewServiceService(nil, nil).CreateService(service)
		if !errors.Is(err, ErrServiceBehaviorInvalid) {
			t.Fatalf("CreateService() error = %v, want %v", err, ErrServiceBehaviorInvalid)
		}
	})

	t.Run("update", func(t *testing.T) {
		repo := &serviceBehaviorTestRepo{
			byID: &models.Service{ID: "service-1", UnitID: "unit-1", Name: "Service"},
		}
		service := &models.Service{ID: "service-1", Behavior: invalid}
		err := NewServiceService(repo, nil).UpdateService(service)
		if !errors.Is(err, ErrServiceBehaviorInvalid) {
			t.Fatalf("UpdateService() error = %v, want %v", err, ErrServiceBehaviorInvalid)
		}
		if repo.updated {
			t.Fatal("UpdateService() persisted an invalid behavior")
		}
	})
}

type serviceBehaviorTestRepo struct {
	byID    *models.Service
	updated bool
}

func (r *serviceBehaviorTestRepo) Create(*models.Service) error { return nil }
func (r *serviceBehaviorTestRepo) CreateTx(*gorm.DB, *models.Service) error {
	return nil
}
func (r *serviceBehaviorTestRepo) NextSortOrderForUnit(string) (int, error) { return 0, nil }
func (r *serviceBehaviorTestRepo) NextSortOrderForUnitTx(*gorm.DB, string) (int, error) {
	return 0, nil
}
func (r *serviceBehaviorTestRepo) FindAllByUnit(string) ([]models.Service, error) { return nil, nil }
func (r *serviceBehaviorTestRepo) FindAllByUnitSubtree(string) ([]models.Service, error) {
	return nil, nil
}
func (r *serviceBehaviorTestRepo) FindByID(string) (*models.Service, error) { return r.byID, nil }
func (r *serviceBehaviorTestRepo) FindByIDTx(*gorm.DB, string) (*models.Service, error) {
	return r.byID, nil
}
func (r *serviceBehaviorTestRepo) FindMapByIDs([]string) (map[string]*models.Service, error) {
	return nil, nil
}
func (r *serviceBehaviorTestRepo) CountByUnitAndIDs(string, []string) (int64, error) {
	return 0, nil
}
func (r *serviceBehaviorTestRepo) CountByUnitSubtreeAndIDs(string, []string) (int64, error) {
	return 0, nil
}
func (r *serviceBehaviorTestRepo) CountByUnitAndCalendarSlotKey(string, string, string) (int64, error) {
	return 0, nil
}
func (r *serviceBehaviorTestRepo) Update(*models.Service) error {
	r.updated = true
	return nil
}
func (r *serviceBehaviorTestRepo) Delete(string) error { return nil }
